src/shipchain_common/utils.py

```diff
@@ -0,0 +1,199 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import decimal
+import json
+import re
+
+from dateutil.parser import parse
+from django.db import models
+from django.conf import settings
+from django.core.mail import EmailMessage
+from django.template.loader import render_to_string
+from enumfields.drf import EnumField
+from rest_framework.exceptions import ValidationError
+
+
+DN_REGEX = re.compile(r'(?:/?)(.+?)(?:=)([^/]+)')
+
+
+def assertDeepAlmostEqual(test_case, expected, actual, *args, **kwargs):  # nopep8 pylint: disable=invalid-name
+    """
+    Assert that two complex structures have almost equal contents.
+
+    Compares lists, dicts and tuples recursively. Checks numeric values
+    using test_case's :py:meth:`unittest.TestCase.assertAlmostEqual` and
+    checks all other values with :py:meth:`unittest.TestCase.assertEqual`.
+    Accepts additional positional and keyword arguments and pass those
+    intact to assertAlmostEqual() (that's how you specify comparison
+    precision).
+
+    :param test_case: TestCase object on which we can call all of the basic
+    'assert' methods.
+    :type test_case: :py:class:`unittest.TestCase` object
+    """
+    is_root = '__trace' not in kwargs
+    trace = kwargs.pop('__trace', 'ROOT')
+    try:
+        if isinstance(expected, (int, float, int, complex)):
+            test_case.assertAlmostEqual(expected, actual, *args, **kwargs)
+        elif isinstance(expected, dict):
+            test_case.assertEqual(set(expected), set(actual))
+            for key in expected:
+                assertDeepAlmostEqual(test_case, expected[key], actual[key],
+                                      __trace=repr(key), *args, **kwargs)
+        else:
+            test_case.assertEqual(expected, actual)
+    except AssertionError as exc:
+        exc.__dict__.setdefault('traces', []).append(trace)
+        if is_root:
+            trace = ' -> '.join(reversed(exc.traces))
+            exc = AssertionError("%s\nTRACE: %s" % (str(exc), trace))
+        raise exc
+
+
+def build_auth_headers_from_request(request):
+    if not request.auth or not isinstance(request.auth, bytes):
+        raise Exception("No auth in request")
+
+    token = request.auth.decode('utf-8')
+    return {'Authorization': f"JWT {token}"}
+
+
+def get_client_ip(request):
+    """
+    Returns ip address from which the http request originated
+    """
+    x_forwarded_for = request.META.get('HTTP_X_FORWARDED_FOR')
+    if x_forwarded_for:
+        addr = x_forwarded_for.split(',')[0]
+    else:
+        addr = request.META.get('REMOTE_ADDR')
+    return addr
+
+
+def get_custom_pagination_class():
+    from rest_framework_json_api.pagination import PageNumberPagination
+
+    class CustomResponsePagination(PageNumberPagination):
```

Comment:
I'm not sure I understand what's going on here, can you show me what this is working around?

Reply:
On `profiles` when `shipchain_common.utils` is imported, the import in line 89 triggers a `djangorestframework-simplejwt` configuration error that I wasn't able to get around other than wrapping this class in a function this way.

Reply:
Oh - I think this could actually be solved by moving the `from rest_framework_json_api.pagination import PageNumberPagination` line to be within `get_paginated_response`. Like:
```
    class CustomResponsePagination(PageNumberPagination):
        def get_paginated_response(self, data):
            from rest_framework_json_api.pagination import PageNumberPagination
            response = super(CustomResponsePagination, self).get_paginated_response(data)
            response.data['data'] = response.data.pop('results')
            return response
```

Give that a shot. There's probably other ways to work around this import issue if that doesn't work.

Reply:
I will give a shot but I'm not sure it will work since `PageNumberPagination` needs to be imported beforehand and it's that import that trigger the issue.

Reply:
@ClovisDj sorry, just getting around to reviewing this again. I am confused by this statement:
```On profiles when shipchain_common.utils is imported, the import in line 89 triggers a djangorestframework-simplejwt configuration error that I wasn't able to get around other than wrapping this class in a function this way.```

As far as I can tell, pagination does not have anything to do with simplejwt? Can you please let me know the exact error you were facing so we can address it?

Also - one other thing I noticed, PageNumberPagination should be imported from rest_framework, not rest_framework_json_api.

Reply:
Right but `rest_framework_json_api` which what we use for pagination [rest_framework_json_api](url) classes somehow imports some module in that package.
I think one possible solution will be just to copy over that root class implementation from `rest_framework_json_api`  in our package.